[WIP] Intuitive manual position control by mapping stick input to acceleration

MaEtUgR opened the pull request:
**Describe problem solved by the proposed pull request**
This is a first draft for a more intuitive manual position controlled flight by stick. Mapping the stick input to velocity generally results in a too agressive flight behavior so often tight acceleration limits have to be applied which result in a too long braking distance and a percieved delay in executing stick inputs.

**Describe your preferred solution**
Mapping the stick inputs to an acceleration looks promising to solve a big part of the listed problems. It feels more intuitive and guarantees gentle vehicle behavior when being gentle on the sticks while allowing to brake hard in emergency cases by pulling the stick all the way to the oposite side of flight direction.

This task requires changes to the position controller to track an
acceleration setpoint which me and @bresch are working on.

**Test data / coverage**
This is a draft and not yet ready to fly on a real vehicle. I'm currently testing in SITL with a joystick to see what changes to the position controller make sense to execute these setpoints.

**Additional context**

* @bresch 's related acceleration feed-forward pr: #12052
* First messy branch I used to thinker around: https://github.com/PX4/Firmware/tree/new-manual-position

resolves #12999

LorenzMeier commented:
Awesome!

hamishwillee commented:
This is just Position mode MC? When this goes in, can you remind me that we will need to update https://docs.px4.io/master/en/flight_modes/position_mc.html (and a few other pages)

Pedro-Roque commented:
This sounds good! But on the other hand we need a good estimation of the accelaration. I can try to help with this :)

xdwgood commented:
Looking forward to its merger, in fact many customers want this feature...Better control feel, better PX4 :smiley:

MaEtUgR commented:
> This is just Position mode MC?

@hamishwillee Yes, it's thought as a replacement option for the current implementation of the manual position mode. At first it will for sure not be default. I will inform you about the behavior and configuration options. Thanks for thinking ahead 👍 

> This sounds good! But on the other hand we need a good estimation of the acceleration.

@Pedro-Roque Certainly would be good to have a useful acceleration estimate but this pr produces acceleration setpoints to be used in a feed-forward manner for the attitude setpoint generation and is not dependent on a good acceleration estimate.

> many customers want this feature...Better control feel, better PX4

@xdwgood Exactly what I've seen as well. Thanks for the confirmation from your side.

Pedro-Roque commented:
> > This is just Position mode MC?
> 
> @hamishwillee Yes, it's thought as a replacement option for the current implementation of the manual position mode. At first it will for sure not be default. I will inform you about the behavior and configuration options. Thanks for thinking ahead 👍
> 
> > This sounds good! But on the other hand we need a good estimation of the acceleration.
> 
> @Pedro-Roque Certainly would be good to have a useful acceleration estimate but this pr produces acceleration setpoints to be used in a feed-forward manner for the attitude setpoint generation and is not dependent on a good acceleration estimate.
> 
> > many customers want this feature...Better control feel, better PX4
> 
> @xdwgood Exactly what I've seen as well. Thanks for the confirmation from your side.

Right, I understood now how you do it after going a bit through the code. Sounds good! Will keep an eye on it

jiejiezhang commented:
@MaEtUgR  Awesome. :+1:

Stifael commented:
Just a comment:
It makes sense to add a new Flighttask that handles acceleration. But this PR is a refactor + new changes and hence the diff is quite large. 
Would it not make sense to add a new task first (position control with acceleration feedforward) and then another PR for refactor?

MaEtUgR commented:
> Would it not make sense to add a new task first (position control with acceleration feedforward) and then another PR for refactor?

Not really because the new task is not even testable without changing the position controller to properly execute acceleration setpoints and introducing this without refactoring would be a nightmare. There were far too many additional corner cases to forget about and have the drone fall down randomly with the old logic...

Stifael commented:
> drone fall down randomly with the old logic

Sure, but then why is it not possible to do first the refactor and then add the new task?

MaEtUgR commented:
I
- rebased on latest master
- added an experimental altitude and position lock implementation for the new task based on integrating the acceleration setpoint to a stop
- added tilt limiting to the acceleration setpoint execution

The main thing missing to finish the acceleration based mapping task is a jerk limit. On racers the instant reactions feel very responsive but it's neither beginner friendly nor suitable for large vehicles.

The pull request should not break any existing functionality and the new task already work just without jerk limit.

MaEtUgR commented:
Seems like
- v2 ran out of flash
- offboard test fails for some reason

MaEtUgR commented:
I finally found the time to rebase and am looking into the failing offboard test that will probably show up in CI again.

MaEtUgR commented:
I rebased and fixed the incompatibilities, we can definitely squash merge this pr in the end since the independent commits are useful for documentation of intent but don't compile anymore. I also found the issues that CI discovered. It's about maximum tilt limitation, looking into a fix.

MaEtUgR commented:
I found why the tilt limit wasn't working and fixed it.

MaEtUgR commented:
Status update: I continued on this by adding unit tests for the `PositionControl` instead of wasting time to just run SITL extensively trying to get the simulated drone into a corner case, going out for a flight and afterwards hope it's all fine. The basic tests for the completely acceleration-based controller (not commited yet) are fine but in the corner cases lies the performance. Some cases are handeled better than before, some worse. Currently I'm fixing the worse ones.

MaEtUgR commented:
I'm getting closer and closer to having this ready. Still a lot of testing needs to be done. Planning the outdoor test of all the pieces together tomorrow.

dagar commented on src/modules/mc_pos_control/PositionControl/ControlMath.hpp:
```diff
@@ -43,6 +43,8 @@
 #include <matrix/matrix/math.hpp>
 #include <uORB/topics/vehicle_attitude_setpoint.h>
 
+static constexpr float A_GRAVITY = 9.80665f; // m/s^2 as defined by the SI standard
```
CONSTANTS_ONE_G

MaEtUgR replied:
Good reference, I didn't know. I'll put that in but it's a bit starnge that then the multicopter position control depends on the ecl/geo library...

MaEtUgR replied:
That's now in the rebased version.

dagar replied:
Which part of that do you find strange? We also depend on it for the projection functions.

MaEtUgR commented:
I had to squash the history because it was not useful at all anymore when rebased.

MaEtUgR commented:
I went test flying yesterday evening, it mostly worked but there were some unsafe corner cases when landing. I fixed them now in the last commits and am looking into the CI catch now. I'll do more test flying today.

MaEtUgR commented:
@PX4/testflights It would be awesome if you could test this pr in all altitude and position-controlled multicopter modes like altitude, position, takeoff, mission, land. 🚁 It contains a lot of changes to the position controller and I'm mostly interested in if you find any regressions. There should only be very small improvements to the flight behavior, I need to do some further work to get the acceleration task to full usability. But this pr gets too big and too old to bundle more things in.

MaEtUgR commented:
I rebased on master without any conflicts to allow testing with the new submodule versions.

jorge789 commented:
### Tested on PixRacer V4:

 **- Modes Tested**
Position Mode: In the second log.Vehicle increased altitude while hovering and with throttle stick in idle position; tested vehicle position by moving forward and backwards a couple times as well as right to left. (Vehicle did not stayed over the trajectory line).
Altitude Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
Stabilized Mode: Good.
Mission Plan Mode (Automated):
RTL (Return To Land): Good, slower than usual.
**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL and see landing behavior.
**- Notes**
Position and Altitude mode have vehicle ascending while hovering.

**Log:**
https://review.px4.io/plot_app?log=77f15ff5-481b-4733-9d19-152700284e31

https://review.px4.io/plot_app?log=de54b3fa-e86c-43c5-8b29-cf57cc62567f

https://review.px4.io/plot_app?log=184e55bd-ef88-46b2-a306-9217bd7e513e

Junkim3DR commented:
### Tested on Pixhawk Pro v4Pro:

**Modes Tested**
- Position Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
- Altitude Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Vehicle follower waypoints but went out of the expected trajectory.
- RTL (Return To Land): Good, slower than usual.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL and see landing behavior.

**Notes**
Position and Altitude mode have vehicle ascending while hovering.

**Log:**
1 - https://review.px4.io/plot_app?log=675855f1-2a37-42c1-aa52-bbf0ab84482a
2 - https://review.px4.io/plot_app?log=69fab381-8a79-48a0-abd4-7385f1588d90



### Tested on Pixhawk 4mini v5:

**Modes Tested**
- Position Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
- Altitude Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Vehicle follower waypoints but went out of the expected trajectory.
- RTL (Return To Land): Good, slower than usual.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL and see landing behavior.

**Notes**
Position and Altitude mode have vehicle ascending while hovering.

**Log:**
1 - https://review.px4.io/plot_app?log=6243b7f7-ab03-4915-9405-ebb66ad2f36e
2 - https://review.px4.io/plot_app?log=e68aec51-7e90-4ecc-8798-e1dabd45b82c



### Tested on NXP FMUK66 v3:

**Modes Tested**
- Position Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
- Altitude Mode: Vehicle increased altitude while hovering and with throttle stick in idle position.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Vehicle follower waypoints but went out of the expected trajectory.
- RTL (Return To Land): Good, slower than usual.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL and see landing behavior.

**Notes**
Position and Altitude mode have vehicle ascending while hovering.

**Log:**
https://review.px4.io/plot_app?log=dec7e76c-1157-4bdf-9130-fef528c185a8

MaEtUgR commented:
Thanks a lot for all the test flights and descriptions! I'll have to investigate using the logs to see if the described notes are caused by my changes. Also my own pr #13130 will generate a small conflict that I might need to resolve.

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -0,0 +1,241 @@
+{
+	Vector3f vel_sp_position = (_pos_sp - _pos).emult(_gain_pos_p);
+	_addIfNotNanVector(_vel_sp, vel_sp_position);
+	_addIfNotNanVector(vel_sp_position, Vector3f());
```
moving into a wrapping function `_setZeroIfNanVector(vel_sp_position)` that contains `_addIfNotNanVector(vel_sp_position, Vector3f());` 

also, maybe invert the two calls such that we clearly see that `_vel_sp` is set to zero if is is NAN and `vel_sp_position` is also NAN

MaEtUgR replied:
> moving into a wrapping function `_setZeroIfNanVector`

Very good idea, makes it more clear. Fixed in https://github.com/PX4/Firmware/pull/12072/commits/ede46989512854320f290e1ffaf492d244324c0e

> also, maybe invert the two calls such that we clearly see that _vel_sp is set to zero if is is NAN and vel_sp_position is also NAN

They cannot be reordered without breaking the functionality because if `vel_sp_position` and `_vel_sp` are both `NAN` there's no velocity setpoint to execute. It's not replaced with zero in that case which allows executing an acceleration setpoint exclusively. If there is no acceleration setpoint the `PositionControl` class will report an error https://github.com/PX4/Firmware/blob/ede46989512854320f290e1ffaf492d244324c0e/src/modules/mc_pos_control/PositionControl/PositionControl.cpp#L114-L117.

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -0,0 +1,241 @@
+	// Constrain horizontal velocity by prioritizing the velocity component along the
+	// the desired position setpoint over the feed-forward term.
+	const Vector2f vel_sp_xy = ControlMath::constrainXY(Vector2f(vel_sp_position),
+				   Vector2f(_vel_sp - vel_sp_position), _constraints.speed_xy);
```
`_param_mpc_xy_vel_max.get()` instead of `_constraints.speed_xy`

MaEtUgR replied:
That's a really good catch, I'll fix that and do a follow up pr that completely removes ambiguous `_constraints`.

MaEtUgR replied:
Fixed in https://github.com/PX4/Firmware/pull/12072/commits/041790132e9388af3745c941e910dc3911371f3f

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -0,0 +1,241 @@
+
+	// Apply Anti-Windup in vertical direction
+	if ((_thr_sp(2) >= -_lim_thr_min && vel_error(2) >= 0.0f) ||
+	    (_thr_sp(2) <= -_lim_thr_max && vel_error(2) <= 0.0f)) {
```
I liked the intermediate variables that were before: `u_max = -_lim_thr_min` to then check `if (_thr_sp(2) >= u_max)`

MaEtUgR replied:
That comment made me rethink how it's implemented and you're right it's confusing to check `_thr_sp(2) >= -_lim_thr_min` and it's only necessary because the parameters are in scalar values and all the vectors including `_thr_sp` are in NED frame. And such cases are exactly what setters are for so I fixed it by converting from [0,1] to [-1,0] in the setter and commented there so it's all more clear.  Fixed in https://github.com/PX4/Firmware/pull/12072/commits/3ecdfbe3775d870c976b0953bee75c99e7a07d72

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -0,0 +1,241 @@
+	}
+
+	// Saturate thrust setpoint in vertical direction
+	_thr_sp(2) = math::constrain(_thr_sp(2), -_lim_thr_max, -_lim_thr_min);
```
also constrain between `u_min` and `u_max`

MaEtUgR replied:
here it can just be used since the setter converted already

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -0,0 +1,241 @@
+	// Update integral part of velocity control
+	_vel_int += vel_error * _gain_vel_i * dt;
+	// Make sure integral doesn't stay NAN
+	_addIfNotNanVector(_vel_int, Vector3f());
```
use `_setZeroIfNanVector(...)`

MaEtUgR replied:
✔️

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -0,0 +1,241 @@
+	const bool valid_acceleration_setpoint = PX4_ISFINITE(_acc_sp(0))
+			&& PX4_ISFINITE(_acc_sp(1))
+			&& PX4_ISFINITE(_acc_sp(2));
+	return valid_acceleration_setpoint;
```
maybe rename to `controller_output_is_valid`. It took me a few seconds to understand why we checked the validity of the acceleration setpoints here. Also, the output is  `_thr_sp`; why don't we check the validity of this vector instead?

MaEtUgR replied:
The output getting executed is actually the attitude setpoint.

> It took me a few seconds to understand why we checked the validity of the acceleration setpoints here.

It's because any input setpoints and state combinations get processed but if at the end there was no valid acceleration vector there has to be at least one dimension in which the inputs were screwed up and the controller cannot generate valid output so this is an emergency.

bresch commented:
@PX4/testflights You reported that the vehicles were ascending during altitude mode. After checking a few logs, it seems that your hover thrust parameter `MPC_THR_HOVER` is set to default (0.5) and that your vehicle hovers at 0.4. Also, the stick isn't centered:
![hover_thr_too_high](https://user-images.githubusercontent.com/14822839/66467298-86d1f480-ea84-11e9-9e13-e6f04b456e1f.png)

bresch commented on src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp:
```diff
@@ -0,0 +1,151 @@
+	_acceleration_setpoint -= cv.emult(_velocity);
+
+	lockAltitude();
+	lockPosition(stick_xy.length());
```
`lockAltitude` knows that it has to take `fabsf(_stick_expo(2))` but you pass `stick_xy.length()` to `lockPosition`. Since those functions are quite similar, they should also get similar parameters.

bresch commented on src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp:
```diff
@@ -0,0 +1,151 @@
+void FlightTaskManualAcceleration::lockPosition(const float stick_input_xy)
+{
+	if (stick_input_xy > FLT_EPSILON) {
+		_position_setpoint(0) = _position_setpoint(1) = NAN;
```
`_position_setpoint.xy() = Vector2f(NAN, NAN)` ? or does it create too much overhead?

bresch replied:
or create the `position_xy` and `velocity_xy` before the `if`, set it to NAN here and move the `position_xy -> _position_setpoint` of lines 138-141 outside of the `if`

bresch commented on src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp:
```diff
@@ -0,0 +1,151 @@
+		_position_setpoint(0) = position_xy(0);
+		_position_setpoint(1) = position_xy(1);
+		_velocity_setpoint(0) = velocity_xy(0);
+		_velocity_setpoint(1) = velocity_xy(1);
```
`_position_setpoint.xy() = position_xy`
`_velocity_setpoint.xy() = velocity_xy`

bresch commented on src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp:
```diff
@@ -0,0 +1,151 @@
+			velocity_xy = Vector2f(_velocity);
+		}
+
+		position_xy += Vector2f(velocity_xy) * _deltatime;
```
`velocity_xy` is already a `Vector2f`

bresch commented on src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp:
```diff
@@ -0,0 +1,151 @@
+		position_xy += Vector2f(velocity_xy) * _deltatime;
+
+		const Vector2f velocity_xy_last = velocity_xy;
+		velocity_xy += Vector2f(_acceleration_setpoint) * _deltatime;
```
```suggestion
		velocity_xy += _acceleration_setpoint.xy() * _deltatime;
```

MaEtUgR commented on src/lib/FlightTasks/tasks/FlightTask/FlightTask.hpp:
```diff
@@ -164,12 +164,12 @@ class FlightTask : public ModuleParams
 
 	void updateVelocityControllerIO(const matrix::Vector3f &vel_sp,
-					const matrix::Vector3f &thrust_sp) {_velocity_setpoint_feedback = vel_sp; _thrust_setpoint_feedback = thrust_sp; }
+					const matrix::Vector3f &thrust_sp) { _velocity_setpoint_feedback = vel_sp; }
```
Yes, I directly wanted to replace it with acceleration feedback to the flight task and already started but didn't want to pack even more functionality into this pr. It wasn't used so far and is still not right now but will definitely get replaced next.

MaEtUgR commented on src/modules/mc_pos_control/PositionControl/PositionControl.hpp:
```diff
@@ -0,0 +1,218 @@
+	 * Set the integral term in xy to 0.
+	 * @see _vel_int
+	 */
+	void resetIntegral() { _vel_int = matrix::Vector3f(); }
```
```suggestion
	void resetIntegral() { _vel_int.zero(); }
```

MaEtUgR commented on src/modules/mc_pos_control/PositionControl/PositionControl.hpp:
```diff
@@ -0,0 +1,218 @@
+	bool update(const float dt);
+
+	/**
+	 * Set the integral term in xy to 0.
```
```suggestion
	 * Set the integral term to 0 to prevent windup.
```

jorge789 commented:
### Tested on PixRacer V4:

**Modes Tested**

Position Mode: 
Altitude Mode: 
Stabilized Mode: 
Mission Plan Mode (Automated): 
RTL: 

**- Procedure**
The parameter MPC_THR_HOVER was changed from 40 to 50. When the vehicle flew, it continued to rise, having the throttle in position and altitude between the two. He wanted to fall and I managed to regain control and land it manually.

Log:
https://review.px4.io/plot_app?log=5449c0f5-8476-4080-9371-f352c6fbef48

Log: RTL BAD
https://review.px4.io/plot_app?log=7d27355a-d70b-474e-84c9-f3a4b309bf6f

MaEtUgR commented:
After rebasing unsuccessfully and realizing that it will probably not be mergeable anytime soon because of risk I'm breaking this down into small pieces.
The first part that should be easy to review is here: https://github.com/PX4/Firmware/pull/13247

MaEtUgR commented:
After rebasing unsuccessfully and realizing that it will probably not be mergeable anytime soon because of risk I'm breaking this down into small pieces.
The first part that should be easy to review is here: https://github.com/PX4/Firmware/pull/13247

LorenzMeier commented:
I'm closing this.
